src/frontend/src/flows/register/captcha.ts

```diff
@@ -1,180 +1,225 @@
-        >
-          Next
-        </button>
-      </div>
-    </form>
-  </div>
-`;
-
-export const confirmRegister = (
-  connection: Connection,
-  captcha: Promise<Challenge>,
-  identity: IdentifiableIdentity,
-  alias: string
-): Promise<LoginFlowResult> => {
-  const container = document.getElementById("pageContent") as HTMLElement;
-  render(pageContent, container);
-  return init(connection, identity, alias, captcha);
-};
-
-const tryRegister = (
-  connection: Connection,
-  identity: IdentifiableIdentity,
-  alias: string,
-  challengeResult: ChallengeResult,
-  func: (result: LoginFlowResult) => void
-) => {
-  withLoader(async () => {
-    return connection.register(identity, alias, challengeResult);
-  }).then((registerResult) => {
-    if (registerResult.kind == "badChallenge") {
-      const confirmParagraph = document.querySelector(
-        ".confirm-paragraph"
-      ) as HTMLElement;
-      confirmParagraph.innerHTML =
-        "The value you entered is incorrect. A new challenge is generated.";
-      requestCaptcha(connection);
-    } else {
-      func(apiResultToLoginFlowResult(registerResult));
+// A symbol that we can differentiate from generic `T` types
+// when verifying the challenge
+export const badChallenge: unique symbol = Symbol("ii.bad_challenge");
+
+export const promptCaptchaTemplate = <T>(props: {
+  cancel: () => void;
+  requestChallenge: () => Promise<Challenge>;
+  verifyChallengeChars: (cr: {
+    chars: string;
+    challenge: Challenge;
+  }) => Promise<T | typeof badChallenge>;
+  onContinue: (result: T) => void;
+}) => {
+  // We define a few Chans that are used to update the page in a
+  // reactive way; see template returned by this function
+
+  // The text and image shown
+  const text = new Chan<TemplateResult>();
+  const img = new Chan<TemplateResult>();
+
+  // The text input where the chars can be typed
+  const input: Ref<HTMLInputElement> = createRef();
+
+  // The "next" button behavior
+  const next = new Chan<((e: SubmitEvent) => void) | undefined>();
+  const nextDisabled = next.map((f) => f === undefined);
+
+  // The "retry" button behavior
+  const retry = new Chan<(() => void) | undefined>();
+  const retryDisabled = retry.map((f) => f === undefined);
+
+  // The various states the component can inhabit
+  type State =
+    | { status: "requesting" }
+    | { status: "prompting"; challenge: Challenge }
+    | { status: "verifying" }
+    | { status: "bad" };
+
+  // On retry, request a new challenge
+  const doRetry = async () => {
+    update({ status: "requesting" });
+    const challenge = await props.requestChallenge();
+    update({ status: "prompting", challenge });
+  };
+
+  // On verification, check the chars and either continue (on good challenge)
+  // or go to "bad" state
+  const doVerify = async (challenge: Challenge) => {
+    update({ status: "verifying" });
+    withRef(input, async (input) => {
+      const res = await props.verifyChallengeChars({
```

Comment:
nth: I would find more readable to deconstruct the props. Here I searched for few seconds what was the "props" and where was that defined since it is not part of the function. If I would have just read the function name, I would have spared these seconds.